File: sdk/npm/src/providers/mail-td.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";
import { createHash } from "crypto";

const CHANNEL: Channel = "mail-td";
const BASE_URL = "https://api.mail.td/api";
const DEFAULT_HEADERS: Record<string, string> = {
  Accept: "application/json",
  "Content-Type": "application/json",
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36",
};

/**
 * 检查哈希是否满足指定前导零位数
 * @param hash - SHA-256 哈希结果
 * @param difficulty - 需要的前导零位数
 */
function checkLeadingZeros(hash: Buffer, difficulty: number): boolean {
  const fullBytes = Math.floor(difficulty / 8);
  const remainBits = difficulty % 8;
  for (let i = 0; i < fullBytes; i++) {
    if (hash[i] !== 0) return false;
  }
  if (remainBits > 0 && fullBytes < hash.length) {
    if ((hash[fullBytes] & ((255 << (8 - remainBits)) & 255)) !== 0)
      return false;
  }
  return true;
}

/**
 * 求解 Proof-of-Work
 * 算法: SHA-256(address + timestamp + nonce)，找到使哈希有 difficulty 个前导零位的 nonce
 * @param address - 邮箱地址（小写）
 * @param timestamp - Unix 秒时间戳
 * @param difficulty - 目标前导零位数
 */
function solvePoW(
  address: string,
  timestamp: number,
  difficulty: number,
): string {
  const base = address + timestamp.toString();
  let nonce = 0;
  while (nonce < 100000000) {
    const input = base + nonce.toString();
    const hash = createHash("sha256").update(input).digest();
    if (checkLeadingZeros(hash, difficulty)) {
      return nonce.toString();
    }
    nonce++;
  }
  throw new Error("mail-td: PoW 求解超时");
}

/**
 * 生成随机字符串
 */
function randomString(length: number): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  let result = "";
  for (let i = 0; i < length; i++) {
    result += chars[Math.floor(Math.random() * chars.length)];
  }
  return result;
}

/**
 * 获取可用域名列表
 */
async function getDomains(): Promise<string[]> {
  const response = await fetchWithTimeout(`${BASE_URL}/domains`, {
    method: "GET",
    headers: DEFAULT_HEADERS,
  });
  if (!response.ok) {
    throw new Error(`mail-td: 获取域名失败 ${response.status}`);
  }
  const data = (await response.json()) as {
    domains: Array<{ domain: string; pro_only: boolean }>;
  };
  return data.domains.filter((d) => !d.pro_only).map((d) => d.domain);
}

/**
 * 创建 mail.td 临时邮箱
 * 流程:
 *   1. 获取可用域名
 *   2. 生成随机用户名和密码
 *   3. 求解 PoW（SHA-256 前导零）
 *   4. POST /accounts 创建账户
 *   5. 返回 JWT token（含账户 ID）
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const domains = await getDomains();
  if (domains.length === 0) {
    throw new Error("mail-td: 无可用域名");
  }

  const domain = domains[Math.floor(Math.random() * domains.length)];
  const username = randomString(10);
  const address = `${username}@${domain}`;
  const password = randomString(20);
  const authKey = createHash("sha256").update(password).digest("hex");

  const addressLower = address.toLowerCase().trim();
  let difficulty = 15;
  let powToken: string | undefined;

  for (let retry = 0; retry < 4; retry++) {
    const timestamp = Math.floor(Date.now() / 1000);
    const nonce = solvePoW(addressLower, timestamp, difficulty);

    const body: Record<string, unknown> = {
      address,
      auth_key: authKey,
      pow: { t: timestamp, n: nonce, d: difficulty },
    };
    if (powToken) {
      (body.pow as Record<string, unknown>).token = powToken;
    }

    const response = await fetchWithTimeout(`${BASE_URL}/accounts`, {
      method: "POST",
      headers: DEFAULT_HEADERS,
      body: JSON.stringify(body),
    });

    const result = (await response.json()) as Record<string, unknown>;

    if (result.status === "retry") {
      difficulty = (result.required_difficulty as number) || difficulty + 2;
      powToken = result.token as string;
      continue;
    }

    if (!response.ok) {
      throw new Error(
        `mail-td: 创建账户失败 ${response.status} ${result.error || ""}`,
      );
    }

    if (!result.address || !result.token) {
      throw new Error("mail-td: 创建账户响应缺少必要字段");
    }

    const token = JSON.stringify({
      jwt: result.token as string,
      id: result.id as string,
    });

    return {
      channel: CHANNEL,
      email: result.address as string,
      token,
    };
  }

  throw new Error("mail-td: PoW 重试次数超限");
}

/**
 * 获取 mail.td 邮件列表
 * @param token - JSON 字符串包含 jwt 和 id
 * @param email - 邮箱地址
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  if (!token) {
    throw new Error("mail-td: 缺少 token");
  }

  let jwt: string;
  let accountId: string;
  try {
    const parsed = JSON.parse(token) as { jwt: string; id: string };
    jwt = parsed.jwt;
    accountId = parsed.id;
  } catch {
    throw new Error("mail-td: token 格式无效");
  }

  const response = await fetchWithTimeout(
    `${BASE_URL}/accounts/${accountId}/messages?page=1`,
    {
      method: "GET",
      headers: {
        ...DEFAULT_HEADERS,
        Authorization: `Bearer ${jwt}`,
      },
    },
  );

  if (!response.ok) {
    throw new Error(`mail-td: 获取邮件失败 ${response.status}`);
  }

  const data = (await response.json()) as { messages: RawMessage[] };
  const messages = Array.isArray(data.messages) ? data.messages : [];

  return messages.map((msg) =>
    normalizeEmail(flattenMessage(msg, email), email),
  );
}

/**
 * 将原始消息转为标准化中间格式
 */
function flattenMessage(
  msg: RawMessage,
  recipient: string,
): Record<string, unknown> {
  return {
    id: msg.id || "",
    from: msg.from?.address || msg.from?.name || "",
    to: recipient,
    subject: msg.subject || "",
    text: msg.text || "",
    html: msg.html || "",
    received_at: msg.created_at || "",
    isRead: msg.seen ?? false,
    attachments: msg.attachments || [],
  };
}

type RawMessage = {
  id?: string;
  from?: { address?: string; name?: string };
  to?: Array<{ address?: string }>;
  subject?: string;
  text?: string;
  html?: string;
  seen?: boolean;
  created_at?: string;
  attachments?: unknown[];
};

File: sdk/go/provider/mail_td.go
package provider

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"strconv"
	"strings"
	"time"

	http "github.com/bogdanfinn/fhttp"
)

/**
 * MailTd — https://api.mail.td
 * 基于 SHA-256 Proof-of-Work 的临时邮箱服务
 * 流程:
 *   1. GET /api/domains 获取可用域名
 *   2. 求解 PoW: SHA-256(address + timestamp + nonce) 需满足 difficulty 个前导零位
 *   3. POST /api/accounts 携带 PoW 创建账户 → 返回 JWT + ID
 *   4. GET /api/accounts/{id}/messages?page=1 携带 Bearer JWT 获取邮件
 * Token 格式: JSON {"jwt":"...","id":"..."}
 */

const mailTdBase = "https://api.mail.td/api"

/**
 * mailTdDomainsResponse — 域名列表响应
 */
type mailTdDomainsResponse struct {
	Domains []struct {
		ID      string `json:"id"`
		Domain  string `json:"domain"`
		ProOnly bool   `json:"pro_only"`
	} `json:"domains"`
}

/**
 * mailTdAccountResponse — 创建账户响应
 */
type mailTdAccountResponse struct {
	ID                      string `json:"id"`
	Address                 string `json:"address"`
	Token                   string `json:"token"`
	Status                  string `json:"status"`
	RequiredDifficulty      int    `json:"required_difficulty"`
	PowToken                string `json:"token_retry"`
	SuggestedNextDifficulty int    `json:"suggested_next_difficulty"`
	Error                   string `json:"error"`
}

/**
 * mailTdMessagesResponse — 邮件列表响应
 */
type mailTdMessagesResponse struct {
	Messages []struct {
		ID   string `json:"id"`
		From struct {
			Address string `json:"address"`
			Name    string `json:"name"`
		} `json:"from"`
		Subject   string `json:"subject"`
		Text      string `json:"text"`
		HTML      string `json:"html"`
		Seen      bool   `json:"seen"`
		CreatedAt string `json:"created_at"`
	} `json:"messages"`
	Page int `json:"page"`
}

/**
 * mailTdTokenData — 内部存储的 token 结构
 */
type mailTdTokenData struct {
	JWT string `json:"jwt"`
	ID  string `json:"id"`
}

/**
 * mailTdCheckLeadingZeros — 检查哈希是否满足 difficulty 个前导零位
 */
func mailTdCheckLeadingZeros(hash [32]byte, difficulty int) bool {
	fullBytes := difficulty / 8
	remainBits := difficulty % 8
	for i := 0; i < fullBytes; i++ {
		if hash[i] != 0 {
			return false
		}
	}
	if remainBits > 0 && fullBytes < 32 {
		mask := byte((255 << (8 - remainBits)) & 255)
		if hash[fullBytes]&mask != 0 {
			return false
		}
	}
	return true
}

/**
 * mailTdSolvePoW — 求解 Proof-of-Work
 * 输入: address + timestamp + nonce
 * 目标: SHA-256 哈希有 difficulty 个前导零位
 */
func mailTdSolvePoW(address string, timestamp int64, difficulty int) (string, error) {
	base := address + strconv.FormatInt(timestamp, 10)
	for nonce := 0; nonce < 100000000; nonce++ {
		input := base + strconv.Itoa(nonce)
		hash := sha256.Sum256([]byte(input))
		if mailTdCheckLeadingZeros(hash, difficulty) {
			return strconv.Itoa(nonce), nil
		}
	}
	return "", fmt.Errorf("mail-td: PoW 求解超时")
}

/**
 * mailTdRandomString — 生成随机小写字母+数字字符串
 */
func mailTdRandomString(n int) string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, n)
	for i := range b {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return string(b)
}

/**
 * MailTdGenerate — 创建 mail.td 临时邮箱
 */
func MailTdGenerate() (*CreatedMailbox, error) {
	client := HTTPClient()

	/* 获取可用域名 */
	req, err := http.NewRequest("GET", mailTdBase+"/domains", nil)
	if err != nil {
		return nil, fmt.Errorf("mail-td: 创建域名请求失败: %w", err)
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", getCurrentUA())

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("mail-td: 域名请求失败: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("mail-td: 读取域名响应失败: %w", err)
	}

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("mail-td: 域名请求 HTTP %d", resp.StatusCode)
	}

	var domsResp mailTdDomainsResponse
	if err := json.Unmarshal(body, &domsResp); err != nil {
		return nil, fmt.Errorf("mail-td: 域名响应解析失败: %w", err)
	}

	/* 筛选免费域名 */
	var freeDomains []string
	for _, d := range domsResp.Domains {
		if !d.ProOnly && d.Domain != "" {
			freeDomains = append(freeDomains, d.Domain)
		}
	}
	if len(freeDomains) == 0 {
		return nil, fmt.Errorf("mail-td: 无可用免费域名")
	}

	domain := freeDomains[rand.Intn(len(freeDomains))]
	username := mailTdRandomString(10)
	address := username + "@" + domain
	password := mailTdRandomString(20)
	authKeyBytes := sha256.Sum256([]byte(password)) //nolint:gosec // mail.td API 认证协议要求，非密码存储
	authKey := fmt.Sprintf("%x", authKeyBytes)

	addressLower := strings.ToLower(strings.TrimSpace(address))
	difficulty := 15
	var powToken string

	/* PoW 重试循环 */
	for retry := 0; retry < 4; retry++ {
		timestamp := time.Now().Unix()
		nonce, err := mailTdSolvePoW(addressLower, timestamp, difficulty)
		if err != nil {
			return nil, err
		}

		powObj := map[string]interface{}{
			"t": timestamp,
			"n": nonce,
			"d": difficulty,
		}
		if powToken != "" {
			powObj["token"] = powToken
		}

		reqBody := map[string]interface{}{
			"address":  address,
			"auth_key": authKey,
			"pow":      powObj,
		}
		reqBytes, _ := json.Marshal(reqBody)

		req2, err := http.NewRequest("POST", mailTdBase+"/accounts", strings.NewReader(string(reqBytes)))
		if err != nil {
			return nil, fmt.Errorf("mail-td: 创建账户请求构建失败: %w", err)
		}
		req2.Header.Set("Content-Type", "application/json")
		req2.Header.Set("Accept", "application/json")
		req2.Header.Set("User-Agent", getCurrentUA())

		resp2, err := client.Do(req2)
		if err != nil {
			return nil, fmt.Errorf("mail-td: 创建账户请求失败: %w", err)
		}

		body2, _ := io.ReadAll(resp2.Body)
		resp2.Body.Close()

		var result map[string]interface{}
		if err := json.Unmarshal(body2, &result); err != nil {
			return nil, fmt.Errorf("mail-td: 创建账户响应解析失败: %w", err)
		}

		/* 检查是否需要增加 difficulty 重试 */
		if status, ok := result["status"].(string); ok && status == "retry" {
			if rd, ok := result["required_difficulty"].(float64); ok {
				difficulty = int(rd)
			} else {
				difficulty += 2
			}
			if tk, ok := result["token"].(string); ok {
				powToken = tk
			}
			continue
		}

		if resp2.StatusCode < 200 || resp2.StatusCode >= 300 {
			errMsg := ""
			if e, ok := result["error"].(string); ok {
				errMsg = e
			}
			return nil, fmt.Errorf("mail-td: 创建账户 HTTP %d: %s", resp2.StatusCode, errMsg)
		}

		resultAddr, _ := result["address"].(string)
		resultToken, _ := result["token"].(string)
		resultID, _ := result["id"].(string)

		if resultAddr == "" || resultToken == "" || resultID == "" {
			return nil, fmt.Errorf("mail-td: 响应缺少必要字段")
		}

		tokenData := mailTdTokenData{JWT: resultToken, ID: resultID}
		tokenBytes, _ := json.Marshal(tokenData)

		return &CreatedMailbox{
			Channel: "mail-td",
			Email:   resultAddr,
			Token:   string(tokenBytes),
		}, nil
	}

	return nil, fmt.Errorf("mail-td: PoW 重试次数超限")
}

/**
 * MailTdGetEmails — 获取 mail.td 邮件列表
 */
func MailTdGetEmails(token, email string) ([]NormEmail, error) {
	if token == "" {
		return nil, fmt.Errorf("mail-td: token 为空")
	}

	var tokenData mailTdTokenData
	if err := json.Unmarshal([]byte(token), &tokenData); err != nil {
		return nil, fmt.Errorf("mail-td: token 格式无效: %w", err)
	}

	if tokenData.JWT == "" || tokenData.ID == "" {
		return nil, fmt.Errorf("mail-td: token 缺少 jwt 或 id")
	}

	client := HTTPClient()
	u := fmt.Sprintf("%s/accounts/%s/messages?page=1", mailTdBase, tokenData.ID)
	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, fmt.Errorf("mail-td: 创建邮件请求失败: %w", err)
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Authorization", "Bearer "+tokenData.JWT)
	req.Header.Set("User-Agent", getCurrentUA())

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("mail-td: 邮件请求失败: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("mail-td: 邮件请求 HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("mail-td: 读取邮件响应失败: %w", err)
	}

	var data mailTdMessagesResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("mail-td: 邮件响应解析失败: %w", err)
	}

	out := make([]NormEmail, 0, len(data.Messages))
	for _, msg := range data.Messages {
		row := map[string]any{
			"id":         msg.ID,
			"from":       msg.From.Address,
			"to":         email,
			"subject":    msg.Subject,
			"text":       msg.Text,
			"html":       msg.HTML,
			"created_at": msg.CreatedAt,
		}
		out = append(out, NormalizeMap(row, email))
	}
	return out, nil
}
